streams/src/test/java/org/apache/kafka/streams/kstream/internals/CogroupedKStreamImplTest.java

```diff
@@ -177,8 +180,603 @@ public void shouldNameProcessorsAndStoreBasedOnNamedParameter() {
+
+        assertThat(
+                topologyDescription,
+                equalTo("Topologies:\n" +
+                        "   Sub-topology: 0\n" +
+                        "    Source: KSTREAM-SOURCE-0000000000 (topics: [one])\n" +
+                        "      --> KSTREAM-MAP-0000000002\n" +
+                        "    Processor: KSTREAM-MAP-0000000002 (stores: [])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter\n" +
+                        "      <-- KSTREAM-SOURCE-0000000000\n" +
+                        "    Processor: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter (stores: [])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-sink\n" +
+                        "      <-- KSTREAM-MAP-0000000002\n" +
+                        "    Sink: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-sink (topic: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition)\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-filter\n\n" +
+                        "  Sub-topology: 1\n" +
+                        "    Source: COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-source (topics: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-0000000007, KSTREAM-AGGREGATE-0000000011\n" +
+                        "    Source: KSTREAM-SOURCE-0000000001 (topics: [two])\n" +
+                        "      --> COGROUPKSTREAM-AGGREGATE-0000000008\n" +
+                        "    Processor: COGROUPKSTREAM-AGGREGATE-0000000007 (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003])\n" +
+                        "      --> COGROUPKSTREAM-MERGE-0000000009\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-source\n" +
+                        "    Processor: COGROUPKSTREAM-AGGREGATE-0000000008 (stores: [COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003])\n" +
+                        "      --> COGROUPKSTREAM-MERGE-0000000009\n" +
+                        "      <-- KSTREAM-SOURCE-0000000001\n" +
+                        "    Processor: COGROUPKSTREAM-MERGE-0000000009 (stores: [])\n" +
+                        "      --> none\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-0000000007, COGROUPKSTREAM-AGGREGATE-0000000008\n" +
+                        "    Processor: KSTREAM-AGGREGATE-0000000011 (stores: [KSTREAM-AGGREGATE-STATE-STORE-0000000010])\n" +
+                        "      --> none\n" +
+                        "      <-- COGROUPKSTREAM-AGGREGATE-STATE-STORE-0000000003-repartition-source\n\n"));
+    }
+
+    @Test
+    public void shouldInsertRepartitionsTopicForUpstreamKeyModificationWithGroupedRemadeWithOptimization() {
```

Comment:
For this test, I'd expect only 2 sub-topologies vs. 3 since `groupedOne` and `groupedFour` have the same key-changing parent.  I'm not sure, I'll have to look into the optimizer code, but I don't think it should hold up the PR though.

Reply:
I think 3 is fine. Note that it's two `map` operators, and the optimizer cannot know that both use the same `Mapper`, ie, each map() could set a different key and thus both cannot be merged.

It would be the same key changing parent if the program would be:
```
final KStream<String, String> stream1 = builder.stream("one", stringConsumed).map((k, v) -> new KeyValue<>(v, k));
final KStream<String, String> stream2 = builder.stream("two", stringConsumed);
final KStream<String, String> stream3 = builder.stream("three", stringConsumed);

final KGroupedStream<String, String> groupedOne = stream1.groupByKey();
final KGroupedStream<String, String> groupedTwo = stream2.groupByKey();
final KGroupedStream<String, String> groupedThree = stream3.groupByKey();
final KGroupedStream<String, String> groupedFour = stream1.groupByKey();
```

Does this make sense?

Reply:
Yep, that's exactly what I wanted to confirm.  After reading your comment I remembered now that it's the reuse of a `KStream` object with the `needsRepartititoning` set to `true` where the optimizer will collapse multiple repartitions